common-lib/src/main/java/com/google/cloud/tools/intellij/util/ThreadUtil.java

```diff
@@ -49,6 +51,22 @@ public static ThreadUtil getInstance() {
   }
 
   /**
+   * Executes the given {@link Runnable} on the event dispatching thread.
+   *
+   * <p>If {@link #invokeLaterOnEDT} is {@code true} the runnable will be invoked asynchronously,
+   * otherwise it will be invoked synchronously.
+   *
+   * @param runnable the {@link Runnable} to run on the EDT.
+   */
+  public void invokeLaterOnEDT(Runnable runnable) {
```

Comment:
Very confused by this method.  It sounds like it's always invoking later, but then it checks a flag in the implementation.  Might be helpful to include use-cases in the javadoc

Reply:
Yeah, I'm not overly crazy about this pattern either. Its used strictly for testing: ONLY unit tests should be replacing the default asynchronous strategy with a synchronous one to aid in testing async code. I was using the same approach as was already used in the ThreadUtil class.

Reply:
I previously had suggested to use to DI to handle these situations: e.g. abstract out the async code into an application service (injected via plugin.xml); then replace it with a mock for unit testing. wdyt?

Reply:
FWIW, this isn't exactly as I suggested. To follow the pattern that already exists in this class (without creating another settable value), you could change this method to simply call:

```
backgroundExecutorService.submit(() -> ApplicationManager.getApplication().invokeAndWait(runnable));
```

Everything else would "take care of itself" as the `backgroundExecutorService` is already swapped for a direct executor in `CloudToolsRule`.

Reply:
ah I see.  Your strategy of letting the runner be swapped for testing sounds good to me.  
@nkibler any thoughts?

Reply:
I think the application service registered via the `plugin.xml` is just extra overhead for relatively the same solution as I listed above. It still modifies a supposedly static object (the `PicoContainer`) in a test utility. The only difference is that there is no `@VisibleForTesting` method exposed, which should fail compilation anyways if production code were to call it.

It's up to you, either way works for me.

Reply:
ah yes. @nkibler - my implementation was a bit silly. I'll refactor to use the already existing executor.

Reply:
I'm OK with this approach since it isolates the smallest component needed for testing, instead of creating a new service which is more overhead (like you said).

Reply:
👍

Reply:
on second look, is:

`backgroundExecutorService.submit(() -> ApplicationManager.getApplication().invokeAndWait(runnable));`

Actually technically equivalent to:

`ApplicationManager.getApplication().invoke(runnable)` ?

That will still block the EDT, no? So in other words, I'm not 100% sure if we can refactor what I have.

Reply:
I'm assuming you meant to type `.invokeAndWait()` in the second statement since there is no `.invoke()`? No they aren't equivalent. The point of using the `backgroundExecutorService` is that it will execute a given `Runnable` using a cached thread pool that is separate from the main thread. `.invokeLater()` technically does this as well, but we don't control the `ExecutorService` that the `Application` uses, so we can't swap it out in unit tests to force it to execute the tasks on the current thread.

Reply:
No I mean invoke() on the second case.

My point was with your suggested refactoring. You said to just replace what I have with:

```
backgroundExecutorService.submit(() -> ApplicationManager.getApplication().invokeAndWait(runnable));
```

What about non-test cases? In this case the background executor will not be a direct one and asynchronously call invokeAndWait. However, invokeAndWait will block the EDT thus not producing the correct results (the UI will block).

Perhaps I'm being stupid and missing something here.

Reply:
oops replace any time I mention `invoke` with `invokeLater`

Reply:
Ah I see. "Invoke and wait" in this sense means to schedule the `Runnable` to run on the EDT and block the current thread until it returns. It won't stop the EDT from running, then force this task to be executed on it before continuing, as that would cause major application issues.

More info from the `.invokeAndWait()` Javadoc:
> Causes `runnable.run()` to be executed synchronously on the AWT event dispatching thread, when the IDE is in the specified modality state (or a state with less modal dialogs open). This call blocks until all pending AWT events have been processed and (then) `runnable.run()` returns.

Reply:
So are we agreeing that we have to keep these boolean setters? :)